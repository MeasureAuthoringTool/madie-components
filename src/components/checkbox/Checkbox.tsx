import React from "react";
import _ from "lodash";
import {
  DisabledDescription,
  DisabledLabel,
  EnabledDescription,
  EnabledLabel,
  Input,
  InputWrapper,
  LabelWrapper,
  OuterWrapper,
} from "./CheckboxStyles";

export interface CheckboxProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  children?: never;
  label: string;
  description?: string;
  disabled?: boolean;
  id?: string;
  checked?: boolean;
  handleChange?: (value: boolean) => void;
  "data-testid"?: string;
}

export default function Checkbox(args: CheckboxProps) {
  const { id, label, description, checked, handleChange, disabled, ...props } =
    args;

  const inputId = id || `checkbox-${_.uniqueId()}`;
  const labelId = `${inputId}-label`;
  const descriptionId = description ? `${inputId}-description` : undefined;

  function changeHandler(e: React.ChangeEvent<HTMLInputElement>) {
    if (handleChange) {
      handleChange(!!e.target.checked);
    }
  }

  const Label = disabled ? DisabledLabel : EnabledLabel;
  const Description = disabled ? DisabledDescription : EnabledDescription;

  return (
    <OuterWrapper>
      <InputWrapper>
        <Input
          id={inputId}
          aria-describedby={descriptionId}
          type="checkbox"
          checked={!!checked}
          onChange={changeHandler}
          disabled={disabled}
          {...props}
        />
      </InputWrapper>
      <LabelWrapper>
        <Label id={labelId} htmlFor={inputId}>
          {label}
        </Label>
        {description ? (
          <Description id={descriptionId}>{description}</Description>
        ) : undefined}
      </LabelWrapper>
    </OuterWrapper>
  );
}
