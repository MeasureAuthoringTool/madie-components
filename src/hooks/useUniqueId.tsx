declare global {
  interface Window {
    idCounter: number;
  }
}

window.idCounter = 0;

export default function useUniqueId(prefix?: string): string {
  window.idCounter = ++window.idCounter || 1;
  return `${prefix || ""}${window.idCounter}`;
}

export function resetUniqueId(): void {
  window.idCounter = 0;
}
