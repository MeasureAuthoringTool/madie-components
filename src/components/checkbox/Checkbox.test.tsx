import * as React from "react";
import { render, fireEvent } from "@testing-library/react";
import { resetUniqueId } from "../../hooks/useUniqueId";
import Checkbox from "./Checkbox";

const testId = "test-checkbox";
const labelId = `${testId}-label`;
const descrId = `${testId}-description`;
const labelText = "Checkbox Label";
const descriptionText = "Checkbox Description";

describe("Checkbox component", () => {
  beforeEach(resetUniqueId);

  it("should render a checkbox component, passing all extra properties to the input element", () => {
    const { getByTestId } = render(
      <Checkbox label={labelText} data-testid={testId} />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    expect(input).toBeInstanceOf(HTMLInputElement);
    expect(input.type).toBe("checkbox");
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
    expect(input.id).toBeDefined();
  });

  it("should render a label for the input", () => {
    const { getByText } = render(<Checkbox label={labelText} id={testId} />);
    const label = getByText(labelText) as HTMLLabelElement;
    expect(label).toBeInstanceOf(HTMLLabelElement);
    expect(label).toBeInTheDocument();
    expect(label).toMatchSnapshot();
    expect(label.id).toBe(labelId);
    expect(label.htmlFor).toBe(testId);
  });

  it("should not render a description if none specified", () => {
    const { container, getByTestId } = render(
      <Checkbox label={labelText} data-testid={testId} id={testId} />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    const description = container.querySelector(`#${descrId}`);
    expect(description).toBeNull();
    expect(input.getAttribute("aria-describedby")).toBeNull();
  });

  it("should render a description if specified", () => {
    const { getByText, getByTestId } = render(
      <Checkbox
        label={labelText}
        data-testid={testId}
        id={testId}
        description={descriptionText}
      />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    const descr = getByText(descriptionText) as HTMLParagraphElement;
    expect(descr).toBeInTheDocument();
    expect(descr).toMatchSnapshot();
    expect(input.getAttribute("aria-describedby")).toBe(descrId);
  });

  it("should use the ID specified", () => {
    const { getByText, getByTestId } = render(
      <Checkbox
        label={labelText}
        data-testid={testId}
        id={testId}
        description={descriptionText}
      />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    const label = getByText(labelText) as HTMLLabelElement;
    const descr = getByText(descriptionText) as HTMLParagraphElement;
    expect(input.id).toBe(testId);
    expect(label.id).toBe(labelId);
    expect(descr.id).toBe(descrId);
  });

  it("should generate an ID if none specified", () => {
    const { getByText, getByTestId } = render(
      <Checkbox
        label={labelText}
        data-testid={testId}
        description={descriptionText}
      />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    const label = getByText(labelText) as HTMLLabelElement;
    const descr = getByText(descriptionText) as HTMLParagraphElement;
    expect(input.id).toBeDefined();
    expect(label.id).toBe(`${input.id}-label`);
    expect(descr.id).toBe(`${input.id}-description`);
  });

  it("should be able to disable", () => {
    const { container, getByTestId } = render(
      <Checkbox
        label={labelText}
        data-testid={testId}
        description={descriptionText}
        disabled={true}
      />
    );
    expect(container).toMatchSnapshot();
    const input = getByTestId(testId) as HTMLInputElement;
    expect(input).toBeDisabled();
  });

  it("should allow specifying a checked state", () => {
    const { getByTestId } = render(
      <Checkbox label={labelText} data-testid={testId} checked={true} />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    expect(input.checked).toBeTruthy();
  });

  it("should trigger the callback when state changes", () => {
    let checkedValue: boolean = false;
    function callback(isChecked: boolean) {
      checkedValue = isChecked;
    }
    const { getByTestId } = render(
      <Checkbox
        label={labelText}
        data-testid={testId}
        checked={false}
        handleChange={callback}
      />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    expect(input.checked).toBeFalsy();
    fireEvent.click(input);
    expect(checkedValue).toBeTruthy();
  });
});
