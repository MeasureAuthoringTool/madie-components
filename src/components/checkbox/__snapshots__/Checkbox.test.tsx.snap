// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should be able to disable 1`] = `
<div>
  <div
    class="sc-iCfMLu hqxfSB"
  >
    <div
      class="sc-furwcr iuZtmL"
    >
      <input
        aria-describedby="checkbox-1-description"
        class="sc-bdvvtL hNbpzH"
        data-testid="test-checkbox"
        disabled=""
        id="checkbox-1"
        type="checkbox"
      />
    </div>
    <div
      class="sc-pVTFL EWVqI"
    >
      <label
        class="sc-gsDKAQ sc-hKwDye iTQrPj gLBWHH"
        for="checkbox-1"
        id="checkbox-1-label"
      >
        Checkbox Label
      </label>
      <p
        class="sc-eCImPb sc-gKclnd hihGzD egGlih"
        id="checkbox-1-description"
      >
        Checkbox Description
      </p>
    </div>
  </div>
</div>
`;

exports[`Checkbox component should render a checkbox component, passing all extra properties to the input element 1`] = `
<input
  class="sc-bdvvtL hNbpzH"
  data-testid="test-checkbox"
  id="checkbox-1"
  type="checkbox"
/>
`;

exports[`Checkbox component should render a description if specified 1`] = `
<p
  class="sc-eCImPb sc-jRQBWg hihGzD bXyiI"
  id="test-checkbox-description"
>
  Checkbox Description
</p>
`;

exports[`Checkbox component should render a label for the input 1`] = `
<label
  class="sc-gsDKAQ sc-dkPtRN iTQrPj BayBM"
  for="test-checkbox"
  id="test-checkbox-label"
>
  Checkbox Label
</label>
`;
